data_scrapers/sonoma.py

```diff
@@ -0,0 +1,299 @@
+        #
+        # new_active = active_cases - cumul_active
+        # active.append({ 'date': date, 'active': new_active, 'cumul_active': active_cases })
+
+    cases.reverse()
+    deaths.reverse()
+    return { 'cases': cases, 'deaths': deaths }
+
+def transform_transmission(transmission_tag: element.Tag) -> Dict[str, int]:
+    """
+    Takes in a BeautifulSoup tag for the transmissions table and breaks it into
+    a dictionary of type:
+    {'community': -1, 'from_contact': -1, 'travel': -1, 'unknown': -1}
+    """
+    transmissions = {}
+    rows = get_rows(transmission_tag)
+    # turns the transmission categories on the page into the ones we're using
+    transmission_type_conversion = {'Community': 'community', 'Close Contact': 'from_contact', 'Travel': 'travel', 'Under Investigation': 'unknown'}
+    for row in rows:
+        type, number, _pct = get_cells(row)
+        if type not in transmission_type_conversion:
+            raise FormatError('The transmission type {0} was not found in transmission_type_conversion'.format(type))
+        type = transmission_type_conversion[type]
+        transmissions[type] = parse_int(number)
+    return transmissions
+
+def transform_tests(tests_tag: element.Tag) -> Dict[str, int]:
+    tests = {}
+    rows = get_rows(tests_tag)
+    for row in rows:
+        result, number, _pct = get_cells(row)
+        lower_res = result.lower()
+        tests[lower_res] = parse_int(number)
+    return tests;
+
+# def generic_transform(tag: element.Tag) -> Dict[str, int]:
+#     """
+#     Transform function for tables which don't require any special processing.
+#     Takes in a BeautifulSoup tag for a table and returns a dictionary
+#     in which the keys are strings and the values integers
+#     """
+#     categories = {}
+#     rows = get_rows(tag)
+#     for row in rows:
+#         cat, cases, _pct = get_cells(row)
+#         categories[cat] = parse_int(cases)
+#     return categories
+
+def gender_transform(tag: element.Tag) -> Dict[str, int]:
+    """
+    Transform function for the cases by gender table.
+    Takes in a BeautifulSoup tag for a table and returns a dictionary
+    in which the keys are strings and the values integers
+    """
+    categories = {}
+    rows = get_rows(tag)
+    gender_string_conversions = {'Males': 'male', 'Females': 'female'}
+    for row in rows:
+        gender, cases, _pct = get_cells(row)
+        if gender not in gender_string_conversions:
+            raise FormatError('An unrecognized gender has been added to the gender table')
+        categories[gender_string_conversions[gender]] = parse_int(cases)
+    return categories
+
+# not sure why I need object in here: I've checked the output value types and
+# they are strings and ints, but mypy is convinced there is an object in there
+def age_transform(tag: element.Tag) -> List[Dict[str, Union[str, int, object]]]:
```

Comment:
Minor nit: should use the same naming convention as the other transform functions, which are all `def transform_xyz`.

> not sure why I need object in here: I've checked the output value types and they are strings and ints, but mypy is convinced there is an object in there

Oh, this is super interesting! I think it has to do with [how Mypy uses contextual info](https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html#context-in-type-inference). For example, this is OK:

```py
def return_union() -> Dict[str, Union[str, int]]:
    a = 'hi'
    b = 5
    return {'a': a, 'b': b}
```

But this causes the problem you’re seeing:

```py
def return_union() -> Dict[str, Union[str, int]]:
    a = 'hi'
    b = 5
    result = {'a': a, 'b': b}
    return result
```

Based on the above docs, I’m all but certain what’s happening is that when Mypy sees:

```py
{'a': a, 'b': b}
# Or in our case, {'group': age_string_transform[group], 'raw_cases': raw_cases}
```

…it has to infer the correct value type. There are a *lot* of valid ways to do that, and what it goes for is the most common super-type of all the values, which is `object` (both `int` and `str` are subtypes of `object`, as is basically everything — it’s pretty similar to `Any`). So the type Mypy infers is `Dict[str, object]`.

BUT when it sees:

```py
return {'a': a, 'b': b}
```

It’s contextually aware enough to know that, because this is being returned, it should go look at the declared return value and see if that would *also* be a compatible alternative. It is, so Mypy just infers that the type of the expression is the one you declared (`Dict[str, Union[str, int]]`).

In the former case, where there wasn’t an immediate return in front of the dict expression, Mypy has no way to tell that it should compare it with the declared return type to see if that’s a better alternative. Then by the time it gets to a return statement, all it knows is that all the dicts you put in the `categories` list are of type `Dict[str, object]`.

**Anyway, long story short, there are two options that can fix this:**

1. When you declare `categories`, declare its type, and append the dict expression to it instead of a variable with the dict (it’s looking at the context again):

    ```py
    categories: List[Dict[str, Union[str, int]]] = []
    ...
        categories.append({'group': age_string_transform[group],
                           'raw_cases': raw_cases})
    ```

2. Declare the dict type before appending it down on line 181:

    ```py
    # Give Mypy a better type to guess than Dict[str, object]
    element: Dict[str, Union[str, int]] = {
        'group': age_string_transform[group],
        'raw_cases': raw_cases
    }
    categories.append(element)
    ```